provider/pkg/googleclient/error_retry_predicates.go

```diff
@@ -0,0 +1,459 @@
+		return true, fmt.Sprintf("reset connection error: %v", err)
+	}
+	return false, ""
+}
+
+// Retry 409s because some APIs like Cloud SQL throw a 409 if concurrent calls
+// are being made.
+//
+//The only way right now to determine it is a retryable 409 due to
+// concurrent calls is to look at the contents of the error message.
+// See https://github.com/hashicorp/terraform-provider-google/issues/3279
+func is409OperationInProgressError(err error) (bool, string) {
+	gerr, ok := err.(*googleapi.Error)
+	if !ok {
+		return false, ""
+	}
+
+	if gerr.Code == 409 && strings.Contains(gerr.Body, "operationInProgress") {
+		logging.V(9).Infof("[DEBUG] Dismissed an error as retryable based on error code 409 and error reason 'operationInProgress': %s", err)
+		return true, "Operation still in progress"
+	}
+	return false, ""
+}
+
+func isSubnetworkUnreadyError(err error) (bool, string) {
+	gerr, ok := err.(*googleapi.Error)
+	if !ok {
+		return false, ""
+	}
+
+	if gerr.Code == 400 && strings.Contains(gerr.Body, "resourceNotReady") && strings.Contains(gerr.Body, "subnetworks") {
+		logging.V(9).Infof("[DEBUG] Dismissed an error as retryable based on error code 400 and error reason 'resourceNotReady' w/ `subnetwork`: %s", err)
+		return true, "Subnetwork not ready"
+	}
+	return false, ""
+}
+
+// GCE (and possibly other APIs) incorrectly return a 403 rather than a 429 on
+// rate limits.
+func is403ReadRequestsForMinuteError(err error) (bool, string) {
+	gerr, ok := err.(*googleapi.Error)
+	if !ok {
+		return false, ""
+	}
+
+	if gerr.Code == 403 && strings.Contains(gerr.Body, "Quota exceeded for quota metric") && strings.Contains(gerr.Body, "Read requests per minute") {
+		logging.V(9).Infof("[DEBUG] Dismissed an error as retryable based on error code 403 and error message 'Quota exceeded for quota metric' on metric `Read requests per minute`: %s", err)
+		return true, "Read requests per minute"
+	}
+	return false, ""
+}
+
+// Retry on comon googleapi error codes for retryable errors.
+// TODO(#5609): This may not need to be applied globally - figure out
+// what retryable error codes apply to which API.
+func isCommonRetryableErrorCode(err error) (bool, string) {
+	gerr, ok := err.(*googleapi.Error)
+	if !ok {
+		return false, ""
+	}
+
+	if gerr.Code == 429 || gerr.Code == 500 || gerr.Code == 502 || gerr.Code == 503 {
+		logging.V(9).Infof("[DEBUG] Dismissed an error as retryable based on error code: %s", err)
+		return true, fmt.Sprintf("Retryable error code %d", gerr.Code)
+	}
+	return false, ""
+}
+
+// We've encountered a few common fingerprint-related strings; if this is one of
+// them, we're confident this is an error due to fingerprints.
+var FINGERPRINT_FAIL_ERRORS = []string{"Invalid fingerprint.", "Supplied fingerprint does not match current metadata fingerprint."}
+
+// Retry the operation if it looks like a fingerprint mismatch.
+func isFingerprintError(err error) (bool, string) {
```

Comment:
Sounds like you can opt-in to retrying on some of these failures dynamically. I am retaining these verbatim for now even if they aren't actually used. Perhaps we could opt-in at the provider level?